service/popm/popm_test.go

```diff
@@ -643,6 +981,136 @@ func TestConnectToBFGAndPerformMine(t *testing.T) {
 	}
 }
 
+func TestConnectToBFGAndPerformMineMultiple(t *testing.T) {
+	privateKey, err := dcrsecp256k1.GeneratePrivateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	server, msgCh, cleanup := createMockBFG(ctx, t, []string{}, false, 2)
+	defer cleanup()
+
+	go func() {
+		miner, err := NewMiner(&Config{
+			BFGWSURL:      server.URL + bfgapi.RouteWebsocketPublic,
+			BTCChainName:  "testnet3",
+			BTCPrivateKey: hex.EncodeToString(privateKey.Serialize()),
+		})
+		if err != nil {
+			panic(err)
+		}
+
+		err = miner.Run(ctx)
+		if err != nil && err != context.Canceled {
+			panic(err)
+		}
+	}()
+
+	// we can't guarantee order here, so test that we get all expected messages
+	// from popm within the timeout
+
+	messagesReceived := make(map[string]int)
+
+	messagesExpected := map[protocol.Command]int{
+		EventConnected:                    1,
+		bfgapi.CmdL2KeystonesRequest:      1,
+		bfgapi.CmdBitcoinInfoRequest:      2,
+		bfgapi.CmdBitcoinBalanceRequest:   2,
+		bfgapi.CmdBitcoinUTXOsRequest:     2,
+		bfgapi.CmdBitcoinBroadcastRequest: 2,
+	}
+
+	for {
+		select {
+		case msg := <-msgCh:
+			t.Logf("received message %v", msg)
+			messagesReceived[msg]++
+		case <-ctx.Done():
+			if ctx.Err() != nil {
+				t.Fatal(ctx.Err())
+			}
+		}
+		missing := false
+		for m := range messagesExpected {
+			message := fmt.Sprintf("%s", m)
+			if messagesReceived[message] != messagesExpected[m] {
+				t.Logf("still missing message %v, found %d want %d", m, messagesReceived[message], messagesExpected[m])
```

Comment:
Will this test pass if there are missing messages? :thinking:

Reply:
@joshuasing it won't pass, if there are missing messages it will time out because it won't break the `for` loop